fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -99,3 +96,131 @@ where
+                TokenAmount::from_atto(BLOCK_GAS_LIMIT),
+            );
+
+            // TODO: In Lotus historical values of the base fee and a more accurate overestimation is performed
+            // for the fee cap. If we issues with messages going through let's consider the historical analysis.
+        }
+
+        Ok(out)
+    }
+
+    // This function performs a simpler implementation of the gas search than the one used in Lotus.
+    // Instead of using historical information of the gas limit for other messages, it searches
+    // for a valid gas limit for the current message in isolation.
+    fn gas_search(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: &mut Message,
+    ) -> anyhow::Result<GasEstimate> {
+        let mut curr_limit = msg.gas_limit;
+
+        while {
+            if let Some(ret) = self.estimation_call_with_limit(state, msg, curr_limit)? {
+                return Ok(ret);
+            }
+
+            curr_limit = (curr_limit as f64 * Self::GAS_SEARCH_STEP) as u64;
+            if curr_limit > BLOCK_GAS_LIMIT {
+                return Ok(GasEstimate {
+                    exit_code: ExitCode::OK,
+                    info: "".to_string(),
+                    gas_limit: BLOCK_GAS_LIMIT,
+                });
+            }
+            true
+        } {}
+
+        // TODO: For a more accurate gas estimation we could track the low and the high
+        // of the search and make higher steps (e.g. `GAS_SEARCH_STEP = 2`).
+        // Once an interval is found of [low, high] for which the message
+        // succeeds, we make a finer-grained within that interval.
+        // At this point, I don't think is worth being that accurate as long as it works.
+
+        Err(anyhow::anyhow!(
+            "gas search failed. no valid gas limit found"
+        ))
+    }
+
+    fn estimation_call_with_limit(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: &Message,
+        limit: u64,
+    ) -> anyhow::Result<Option<GasEstimate>> {
+        let mut msg = msg.clone();
+        msg.gas_limit = limit;
+        // set message nonce to zero so the right one is picked up
+        msg.sequence = 0;
+        println!("message being applied: {:?}", msg);
+
+        let apply_ret = state.call_with_cache(msg, false)?;
+
+        let ret = GasEstimate {
+            exit_code: apply_ret.msg_receipt.exit_code,
+            info: apply_ret
+                .failure_info
+                .map(|x| x.to_string())
+                .unwrap_or_default(),
+            gas_limit: apply_ret.msg_receipt.gas_used,
+        };
+
+        if ret.exit_code == ExitCode::OK {
+            return Ok(Some(ret));
+        }
+
+        if ret.exit_code != ExitCode::SYS_OUT_OF_GAS {
+            return Err(anyhow::anyhow!(
+                "message execution failed in gas search with error code {:?}",
+                ret.exit_code
+            ));
+        }
```

Comment:
So this is what we have to be very careful with, the way you use it will kill the process. 

If I were you I would add an `is_out_of_gas` method to the `GasEstimate` struct and decide on that in `gas_search`. It would simplify the decision logic 
* from `Ok(Some) | Ok(None) | Err` 
* to `Ok(est) | Err` and looking at `est.is_out_of_gas()` which is a bit more readable as well on the caller side.